src/jit/lowerxarch.cpp

```diff
@@ -1174,13 +1174,13 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     else
     {
         MakeSrcContained(tree, shiftBy);
-    }
 
-    // Codegen of shift oper sets ZF and SF flags.
-    // Note that Rotate Left/Right instructions don't set ZF and SF flags.
-    if (tree->OperIsShift())
-    {
-        tree->gtFlags |= GTF_ZSF_SET;
+        // Codegen of shift oper sets ZF and SF flags provided shift count is not zero.
+        // Note that Rotate Left/Right instructions don't set ZF and SF flags.
+        if (tree->OperIsShift() && (shiftBy->gtIntConCommon.IconValue() != 0))
+        {
+            tree->gtFlags |= GTF_ZSF_SET;
+        }
```

Comment:
It looks like you did disable this optimization for variable sized shifts.

It is surprising that there are no asm diffs for this change.